src/module.c

```diff
@@ -1837,6 +1837,28 @@ unsigned long long RM_GetClientId(RedisModuleCtx *ctx) {
     return ctx->client->id;
 }
 
+/* Return the ACL user name used by the client with the specified client ID.
+ * Client ID can be obtained with RM_GetClientId() API. If the client does not
+ * exist, NULL is returned and errno is set to ENOTDIR. If the client isn't 
+ * using an ACL user, NULL is returned and errno is set to ENODEV */
+RedisModuleString *RM_GetClientUserNameById(RedisModuleCtx *ctx, uint64_t id) {
+    client *client = lookupClientByID(id);
+    if (client == NULL) {
+        errno = ENOTDIR;
+        return NULL;
+    }
+    
+    if (client->user == NULL) {
+        errno = ENODEV;
+        return NULL;
+    }
```

Comment:
I like returning NULL but I'm not sure I like the errno values (ENOTDIR = No such directory, ENODEV = No such device).

How about using ENOENT for `client == NULL` (No such entity, where entity can be anything, such as a user) and ENOTSUP for `client->user == NULL` (Operation not supported for the fake user) instead?

Reply:
`ENOENT` and `ENOTSUP` is really realistic!  The reason I choose `ENOTDIR` and `ENODEV` is because  this two code is already in [acl.c](https://github.com/redis/redis/blob/unstable/src/acl.c#L819).  At the same time. Two code expressed the hierarchical relationship just like struct and field relationship .

Reply:
I'm happy you like the idea. :relaxed:

I think it's more important that errno is consistent within the Module API, because it needs to be backward compatible. The rest of Redis code can be changed more easily.

Errno was used extensively in the new stream API, for example RM_StreamDelete: https://redis.io/topics/modules-api-ref#coderedismodulestreamdeletecode.

Reply:
@zuiderkwast  Thanks for you review~~~ . I have created a new commit for changing ERRNO

> How about using ENOENT for client == NULL (No such entity, where entity can be anything, such as a user) and ENOTSUP for client->user == NULL (Operation not supported for the fake user) instead?

@madolson @oranagra  Please take a look 😄  I think zuiderkwast's suggestion is more expressive